functions/imagemagick/main.py

```diff
@@ -0,0 +1,73 @@
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the 'License');
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an 'AS IS' BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+# [START functions_imagemagick_setup]
+import os
+
+from google.cloud import storage, vision
+from wand.image import Image
+
+storage_client = storage.Client()
+vision_client = vision.ImageAnnotatorClient()
+# [END functions_imagemagick_setup]
+
+
+# [START functions_imagemagick_analyze]
+# Blurs uploaded images that are flagged as Adult or Violence.
+def blur_offensive_images(data, context):
+    file_data = data
+
+    file_name = file_data['name']
+    blob = storage_client.bucket(file_data['bucket']).get_blob(file_name)
+    blob_uri = 'gs://%s/%s' % (file_data['bucket'], file_name)
+    blob_source = {'source': {'image_uri': blob_uri}}
+
+    print('Analyzing %s.' % file_name)
+
+    result = vision_client.safe_search_detection(blob_source)
+    detected = result.safe_search_annotation
+
+    if detected.adult == 5 or detected.violence == 5:
+        print('The image %s was detected as inappropriate.' % file_name)
+        return __blur_image(blob)
+    else:
+        print('The image %s was detected as OK.' % file_name)
+# [END functions_imagemagick_analyze]
+
+
+# [START functions_imagemagick_blur]
+# Blurs the given file using ImageMagick.
+def __blur_image(blob):
+    file_name = blob.name
+    temp_local_filename = '/tmp/%s' % os.path.basename(file_name)
```

Comment:
Rather than hardcoding /tmp, can we use `tempfile.mkdtemp` to get a random, guaranteed-temporary directory? (note, we should clean this up afterwards as well)

Reply:
(One reason is to avoid filename conflicts or race conditions.)

Reply:
Done (with `mkstemp` instead.)